File: src/mcconsHeuristic.cpp
// MCCONS using a genetic algorithm solver

#include "../include/MCCONS.h"
#include "../include/SolverGA.h"
#include "../include/OptionParser.h"


using optparse::OptionParser;


int main(int argc, char *argv[])
{

    // create the command line parser
    OptionParser parser = OptionParser().description("MC-Cons Consensus Optimizer Using a Genetic Algorithm");

    // I/O settings
    parser.add_option("-i", "--data").dest("dataFile").help("path to MARNA-like input file");
    parser.add_option("-s", "--silent").action("store_true").dest("silent").help("don't display status to stderr");

    // heuristics parameters
    parser.add_option("-p", "--popsize").dest("popSize").help("genetic algorithm population size").type("size_t");
    parser.add_option("-n", "--numgen").dest("numGenerations").type("size_t").help("genetic algorithm number of generations");

    // threshold on tree consensus
    parser.add_option("-t", "--threshold").dest("threshold").type("size_t").help("consider up to t additional score for tree consensus (vs optiomal one)");

    // prng seeds
    parser.add_option("--seed0").dest("seed0").type("long").help("seed 0 used by the prng");
    parser.add_option("--seed1").dest("seed1").type("long").help("seed 1 used by the prng");
    parser.add_option("--seed2").dest("seed2").type("long").help("seed 2 used by the prng");
    parser.add_option("--seed3").dest("seed3").type("long").help("seed 3 used by the prng");
    parser.add_option("--seed4").dest("seed4").type("long").help("seed 4 used by the prng");
    parser.add_option("--seed5").dest("seed5").type("long").help("seed 5 used by the prng");


//     parser.add_option("--sedd2").dest("seed2").type("unsigned)
    optparse::Values options = parser.parse_args(argc, argv);


    // GLOBAL SETTINGS (MODIFY AT YOUR OWN RISKS (WHICH ARE MINIMAL))
    unsigned long SEEDS[6] = {42, 42, 42, 42, 42, 42};

    if (options.is_set("seed0"))
    {
        SEEDS[0] = atoi(options["seed0"].c_str());
    }
    if (options.is_set("seed1"))
    {
        SEEDS[1] = atoi(options["seed1"].c_str());
    }
    if (options.is_set("seed2"))
    {
        SEEDS[2] = atoi(options["seed2"].c_str());
    }
    if (options.is_set("seed3"))
    {
        SEEDS[3] = atoi(options["seed3"].c_str());
    }
    if (options.is_set("seed4"))
    {
        SEEDS[4] = atoi(options["seed4"].c_str());
    }
    if (options.is_set("seed5"))
    {
        SEEDS[5] = atoi(options["seed5"].c_str());
    }


    // check that the seeds aren't all zeros, because the generator stays in the same
    // state if that happens
    int num_zeros = 0.;
    for (size_t i = 0; i != 6; ++i)
    {
        if (SEEDS[i] == 0)
        {
            num_zeros += 1;
        }
    }
    if (num_zeros == 6)
    {
        std::cerr << "The seeds of the pseudorandom number generator must not be all zeros." << std::endl;
        std::exit(0);
    }

    int IMPROVEMENT_DEPTH= 2;
    int ELITE_SIZE = 30;
    double CROSSOVER_PROBABILITY = 0.5;
    double MUTATION_PROBABILITY = 0.05;
    double IMPROVEMENT_PROBABILITY = 0.05;
    double SUBOPTIMAL_THRESHOLD = 0;
    size_t POPULATION_SIZE = 250;
    size_t NUM_GENERATIONS = 250;
    bool silent = false;


    if (options.is_set("dataFile"))
    {
      // extract command line parameters
      if (options["silent"] == "1")
      {
          silent = true;
      }

      if (options.is_set("popSize"))
      {
          POPULATION_SIZE = atoi(options["popSize"].c_str());
      }

      if (options.is_set("numGenerations"))
      {
          NUM_GENERATIONS = atoi(options["numGenerations"].c_str());
      }

      if (options.is_set("threshold"))
      {
          SUBOPTIMAL_THRESHOLD = atof(options["threshold"].c_str());
      }


      std::string path = options["dataFile"];

      // instantiate the genetic algorithm solver
      // tree solver, will consider suboptimal consensus
      Solver* tree_solver = new SolverGA(silent,
                                         POPULATION_SIZE,
                                         NUM_GENERATIONS,
                                         IMPROVEMENT_DEPTH,
                                         ELITE_SIZE,
                                         SUBOPTIMAL_THRESHOLD, // specified subopt threshold for trees

                                         CROSSOVER_PROBABILITY,
                                         MUTATION_PROBABILITY,
                                         IMPROVEMENT_PROBABILITY);

      Solver* dot_bracket_solver = new SolverGA(silent,
                                                POPULATION_SIZE,
                                                NUM_GENERATIONS,
                                                IMPROVEMENT_DEPTH,
                                                ELITE_SIZE,
                                                0.,  // no threshold for this one

                                                CROSSOVER_PROBABILITY,
                                                MUTATION_PROBABILITY,
                                                IMPROVEMENT_PROBABILITY);

      // execute and cleanup
      MCCONS(path, tree_solver, dot_bracket_solver, SEEDS);
      delete tree_solver;
      delete dot_bracket_solver;

    } else
    {
      // something went wrong with the arguemnts, print error message and exit
      std::cerr << "Error: something went wrong, please check usage -h or --help" << std::endl;
      std::exit(0);
    }
}

File: java/src/MCConsHeuristic.java
import com.martiansoftware.jsap.FlaggedOption;
import com.martiansoftware.jsap.JSAP;
import com.martiansoftware.jsap.JSAPResult;
import com.martiansoftware.jsap.Switch;


public final class MCConsHeuristic {

    public static void main(String[] args) throws Exception {
        JSAP parser = new JSAP("MC-Cons Consensus Optimizer Using A Heuristic Approach");

        // I/O settings
        FlaggedOption input = new FlaggedOption("input", JSAP.STRING_PARSER, null, true,
                'i', "input", "input file path");
        parser.registerParameter(input);

        FlaggedOption output = new FlaggedOption("output", JSAP.STRING_PARSER, "stdout", false,
                'o', "output", "output file destination (default is stdout)");
        parser.registerParameter(output);


        // verbosity
        Switch verbose = new Switch("verbose", 'v', "verbose", "verbosity of the solver");
        parser.registerParameter(verbose);

        // suboptimality tolerance
        FlaggedOption tolerance = new FlaggedOption("tolerance", JSAP.DOUBLE_PARSER, "0.", false,
                't', "tolerance", "Permitted gap between kept solution and best solution, normalized");
        parser.registerParameter(tolerance);

        // random number generator seeds
        FlaggedOption seed0 = new FlaggedOption("seed0", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed0", "First seed of the random stream");
        parser.registerParameter(seed0);
        FlaggedOption seed1 = new FlaggedOption("seed1", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed1", "Second seed of the random stream");
        parser.registerParameter(seed1);
        FlaggedOption seed2 = new FlaggedOption("seed2", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed2", "Third seed of the random stream");
        parser.registerParameter(seed2);
        FlaggedOption seed3 = new FlaggedOption("seed3", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed3", "Fourth seed of the random stream");
        parser.registerParameter(seed3);
        FlaggedOption seed4 = new FlaggedOption("seed4", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed4", "Fifth seed of the random stream");
        parser.registerParameter(seed4);
        FlaggedOption seed5 = new FlaggedOption("seed5", JSAP.LONG_PARSER, "42", false,
                JSAP.NO_SHORTFLAG, "seed5", "Sixth seed of the random stream");
        parser.registerParameter(seed5);


        // heuristic parameters
        FlaggedOption popSize = new FlaggedOption("popSize", JSAP.INTEGER_PARSER, "150", false,
                'p', "popSize", "Genetic algorithm population size");
        parser.registerParameter(popSize);

        FlaggedOption numGenerations = new FlaggedOption("numGenerations", JSAP.INTEGER_PARSER, "150", false,
                'n', "numGenerations", "Number of generations to evaluate");
        parser.registerParameter(numGenerations);

        FlaggedOption eliteSize = new FlaggedOption("eliteSize", JSAP.INTEGER_PARSER, "30", false,
                JSAP.NO_SHORTFLAG, "eliteSize", "Size of the elite in the genetic algorithm");
        parser.registerParameter(eliteSize);

        FlaggedOption improvementProbability = new FlaggedOption("improvementProbability", JSAP.DOUBLE_PARSER, "0.05",
                false, JSAP.NO_SHORTFLAG, "improvementProbability",
                "Probability of applying steepest descent on a solution");
        parser.registerParameter(improvementProbability);

        FlaggedOption improvementDepth = new FlaggedOption("improvementDepth", JSAP.INTEGER_PARSER, "2",
                false, JSAP.NO_SHORTFLAG, "improvementDepth",
                "Number of iterations applied each time the steepest descent is called");
        parser.registerParameter(improvementDepth);

        FlaggedOption crossoverProbability = new FlaggedOption("crossoverProbability", JSAP.DOUBLE_PARSER, "0.5",
                false, JSAP.NO_SHORTFLAG, "crossoverProbability",
                "Probability of applying uniform crossover on a child solution");
        parser.registerParameter(crossoverProbability);

        FlaggedOption crossoverMixingRatio = new FlaggedOption("crossoverMixingRatio", JSAP.DOUBLE_PARSER, "0.5",
                false, JSAP.NO_SHORTFLAG, "crossoverMixingRatio",
                "Mixing ratio used for uniform crossover between two parent solutions");
        parser.registerParameter(crossoverMixingRatio);


        FlaggedOption mutationProbability = new FlaggedOption("mutationProbability", JSAP.DOUBLE_PARSER, "0.05",
                false, JSAP.NO_SHORTFLAG, "mutationProbability",
                "Probability of applying mutation on a child solution");
        parser.registerParameter(mutationProbability);

        FlaggedOption mutationStrength = new FlaggedOption("mutationStrength", JSAP.DOUBLE_PARSER, "0.3",
                false, JSAP.NO_SHORTFLAG, "mutationStrength",
                "Probability that a gene is mutated");
        parser.registerParameter(mutationStrength);


        // parse the arguments
        JSAPResult config = parser.parse(args);


        if (!config.success()) {

            System.err.println();

            // print out specific error messages describing the problems
            // with the command line, THEN print usage, THEN print full
            // help.  This is called "beating the user with a clue stick... lol
            for (java.util.Iterator errs = config.getErrorMessageIterator();
                 errs.hasNext(); ) {
                System.err.println("Error: " + errs.next());
            }

            System.err.println("                "
                    + parser.getUsage());
            System.err.println();
            System.err.println(parser.getHelp());
            System.exit(1);
        }

        long[] SEEDS = {config.getLong("seed0"), config.getLong("seed1"), config.getLong("seed2"),
                config.getLong("seed3"), config.getLong("seed4"), config.getLong("seed5")};


        SolverHeuristic tree_solver = new SolverHeuristic(
                config.getBoolean("verbose"),
                config.getDouble("tolerance"),
                SEEDS,

                config.getInt("popSize"),
                config.getInt("numGenerations"),
                config.getInt("eliteSize"),

                config.getDouble("improvementProbability"),
                config.getInt("improvementDepth"),

                config.getDouble("crossoverProbability"),
                config.getDouble("crossoverMixingRatio"),

                config.getDouble("mutationProbability"),
                config.getDouble("mutationStrength"));


        SolverHeuristic dot_bracket_solver = new SolverHeuristic(
                config.getBoolean("verbose"),
                0.,
                SEEDS,

                config.getInt("popSize"),
                config.getInt("numGenerations"),
                config.getInt("eliteSize"),

                config.getDouble("improvementProbability"),
                config.getInt("improvementDepth"),

                config.getDouble("crossoverProbability"),
                config.getDouble("crossoverMixingRatio"),

                config.getDouble("mutationProbability"),
                config.getDouble("mutationStrength"));

        // execute MC-Cons
        MCCons.MCCONS(config.getString("input"), tree_solver, dot_bracket_solver);
    }
}

